flink-filesystems/flink-s3-fs-base/src/main/java/org/apache/flink/fs/s3/common/writer/S3RecoverableFsDataOutputStream.java

```diff
@@ -126,7 +126,16 @@ public long getPos() throws IOException {
 
     @Override
     public void sync() throws IOException {
-        fileStream.sync();
```

Comment:
`sync` method is called on the following scenerios 
1. `S3RecoverableWriter`
2. `FlinkS3FileSystem` creates new instance of `S3RecoverableWriter` when `createRecoverableWriter()` method is called 
3. `CsvBulkWriter` uses `FlinkS3FileSystem` and calls recoverableWriter. 
4. `BulkWriter`


This change will not alter any processing guarantee.  

In the current changes in `sync()` method , it takes the lock first then makes a call to filesystem flush and  commits remaining blocks (writes to s3). This flow results in exactly once . Same code flow is implemented for `AzureBlobFsRecoverableDataOutputStream` . 

From the class `BlockBlobAppendStream` 

```
    public void hsync() throws IOException {
        if (this.compactionEnabled) {
            this.flush();
        }

    }
```

Reply:
> This change will not alter any processing guarantee.

I can say 100% sure that we're having issue with this now.  The reasons is the following:
* `sync` method is defined in `FSDataOutputStream` with the following definition:
```
    /**
     * Flushes the data all the way to the persistent non-volatile storage (for example disks). The
     * method behaves similar to the <i>fsync</i> function, forcing all data to be persistent on the
     * devices.
     *
     * @throws IOException Thrown if an I/O error occurs
     */
```
* In case `sync` method call user of the writer instance is expected to call further `write` methods
* What is actually happening it's blowing up the next write with the following exception:
```
java.io.IOException: Stream closed.

	at org.apache.flink.core.fs.RefCountedFileWithStream.requireOpened(RefCountedFileWithStream.java:72)
	at org.apache.flink.core.fs.RefCountedFileWithStream.write(RefCountedFileWithStream.java:52)
	at org.apache.flink.core.fs.RefCountedBufferingFileStream.flush(RefCountedBufferingFileStream.java:104)
	at org.apache.flink.core.fs.RefCountedBufferingFileStream.write(RefCountedBufferingFileStream.java:87)
	at org.apache.flink.fs.s3.common.writer.S3RecoverableFsDataOutputStream.write(S3RecoverableFsDataOutputStream.java:112)
	at java.base/java.io.OutputStream.write(OutputStream.java:122)
```
* This can be super easily tested with `S3RecoverableFsDataOutputStreamTest.testSync`. Please remove the `expected = Exception.class` from the beginning of the test.
* The following line in the test is testing nothing because never ever called.

Reply:
@dannycranmer WDYT?